### PR TITLE
EdgeSite working with standard K8s master

### PR DESCRIPTION
**Pull request by CindyXing**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind test
> /kind failing-test

/kind feature

**What this PR does / why we need it**:
This change enables KubeEdge agent talking to a K8s master such as K3S server. 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by m1093782566** [comment]
A quick question, should not we treat edgesite as an independent project?

**Comment by CindyXing** [comment]
> A quick question, should not we treat edgesite as an independent project?

I'd like to make sure we brand the edgesite as part of KubeEdge project. Also edgesite mainly is to build a new client side.

**Comment by rohitsardesai83** [comment]
@CindyXing , the idea of edge site is good . Is there some way we can avoid code duplication ?

**Comment by sids-b** [comment]
> @CindyXing , the idea of edge site is good . Is there some way we can avoid code duplication ?

+1 to @rohitsardesai83 . All our settings are configuration based. I feel having a edgesite.go , conf and few changes in edge modules should do the work.

File: vendor/github.com/kubeedge/beehive/pkg/common/util/parse_resource.go [hunk 1]
```diff
@@ -13,6 +13,8 @@ func ParseResourceEdge(resource string, operation string) (string, string, strin
 	resourceSplits := strings.Split(resource, "/")
 	if len(resourceSplits) == 3 {
 		return resourceSplits[0], resourceSplits[1], resourceSplits[2], nil
+	} else if len(resourceSplits) == 5 {
```
Comment (sids-b) [inline vendor/github.com/kubeedge/beehive/pkg/common/util/parse_resource.go]: We need to change in https://github.com/kubeedge/beehive and then update the vendor.

**Comment by sids-b** [comment]
> > A quick question, should not we treat edgesite as an independent project?
> 
> I'd like to make sure we brand the edgesite as part of KubeEdge project. Also edgesite mainly is to build a new client side.

Agree, kubeedge should be able to support EdgeSite scenarios as well.

**Comment by CindyXing** [comment]
> > @CindyXing , the idea of edge site is good . Is there some way we can avoid code duplication ?
> 
> +1 to @rohitsardesai83 . All our settings are configuration based. I feel having a edgesite.go , conf and few changes in edge modules should do the work. I can help with it :)

Yes, agree that we should prevent dup code. That's why most of the code is referenced from edge folder.  Currently there is some change in the controller/manager that it'll only watch pods bond to the current node. There can be other change(s) needed. So, for now, let's keep separate and then adjust accordingly.

**Comment by sids-b** [comment]
The design doc mentions about lightweight etcd, where does that fit here ? Or will we be using sqlite ?

**Comment by CindyXing** [comment]
> The design doc mentions about lightweight etcd, where does that fit here ? Or will we be using sqlite ?
lightweight etcd is for the control plane not agent. 

> The design doc mentions about lightweight etcd, where does that fit here ? Or will we be using sqlite ?

What do you mean @sids-b? Either sqlite or lightweight etcd. A lightweight etcd can be used when it is available. And it is customer's choice.

**Comment by sids-b** [comment]
Ohh, I get it now, so metaManager uses either sqlite or ligtweight etcd based on configurations of customer. Sounds good :)

File: edgesite/pkg/controller/constants/service.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	ResourceResourceTypeIndex = 1
+	ResourceResourceNameIndex = 2
+	ResourceNode              = "node"
+
+	// Group
+	GroupResource = "resource"
+
+	// Nvidia Constants
+	// NvidiaGPUStatusAnnotationKey is the key of the node annotation for GPU status
+	NvidiaGPUStatusAnnotationKey = "huawei.com/gpu-status"
```
Comment (kevin-wangzefeng) [inline edgesite/pkg/controller/constants/service.go]: Please replace `huawei.com` with `gpu.kubeedge.io`or something else, it has nothing to do with the vendor here.

File: edgesite/pkg/controller/config/kube.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import (
+	"time"
+
+	"github.com/kubeedge/beehive/pkg/common/config"
+	"github.com/kubeedge/beehive/pkg/common/log"
+	"github.com/kubeedge/kubeedge/edgesite/pkg/controller/constants"
+)
+
+// KubeMaster is the url of edge master(kube api server)
+var KubeMaster string
```
Comment (kevin-wangzefeng) [inline edgesite/pkg/controller/config/kube.go]: `KubeAPIEndpoint`?

File: edgesite/pkg/controller/constants/service.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	// module
+	EdgeControllerModuleName     = "controller"
+	CloudHubControllerModuleName = "cloudhub"
+
+	// Resource sep
+	ResourceSep               = "/"
+	ResourceSliceLength       = 5
+	ResourceSliceLengthQuery  = 4
+	ResourceNodeIndex         = 0
+        ResourceNodeIDIndex       = 1
```
Comment (kevin-wangzefeng) [inline edgesite/pkg/controller/constants/service.go]: inaccurate indent

File: edgesite/pkg/controller/controller/upstream.go [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+					continue
+				}
+
+				if err != nil {
+					log.LOGGER.Warnf("message: %s process failure with error: %s, name: %s", msg.GetID(), err, namespace, name)
+					continue
+				}
+
+				// In case the status stored at metadata service is outdated, update the heartbeat automatically
+				/*for i := range nodeStatusRequest.Status.Conditions {
```
Comment (kevin-wangzefeng) [inline edgesite/pkg/controller/controller/upstream.go]: Why it is period commented?
If it's intended, please describe when to uncomment this.

File: edgesite/pkg/controller/config/kube.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+		KubeConfig = kc
+	}
+	log.LOGGER.Infof("kube config: %s", KubeConfig)
+
+	if kn, err := config.CONFIG.GetValue("controller.kube.namespace").ToString(); err != nil {
+		KubeNamespace = constants.DefaultKubeNamespace
+	} else {
+		KubeNamespace = kn
+	}
+	log.LOGGER.Infof("kube namespace: %s", KubeNamespace)
```
Comment (kevin-wangzefeng) [inline edgesite/pkg/controller/config/kube.go]: Where is KubeNamespace used? I can't find it

File: edgesite/pkg/controller/config/kube.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+// NodeID for the current node
+var KubeNodeID string
+
+// NodeName for the current node
+var KubeNodeName string
+
+// KubeUpdateNodeFrequency is the time duration for update node status(default is 20s)
+var KubeUpdateNodeFrequency time.Duration
+
+func init() {
```
Comment (kevin-wangzefeng) [inline edgesite/pkg/controller/config/kube.go]: Alternative to copy field by field, I'd suggest:
1. wrap all `KubeX` fields into a struct
2. wrap all default values into several structs
3. use struct copy / deep copy to set default values for `KubeX` fields, and then rewrite those user indicated ones.

File: edgesite/pkg/controller/config/kube.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		KubeContentType = constants.DefaultKubeContentType
+	} else {
+		KubeContentType = kct
+	}
+	log.LOGGER.Infof("kube content type: %s", KubeContentType)
+
+	if kqps, err := config.CONFIG.GetValue("controller.kube.qps").ToFloat64(); err != nil {
+		KubeQPS = constants.DefaultKubeQPS
+	} else {
+		KubeQPS = float32(kqps)
```
Comment (kevin-wangzefeng) [inline edgesite/pkg/controller/config/kube.go]: It looks confusing that the code check with float64 but set value with float32, maybe define KubeQPS with float64?

Comment (CindyXing) [reply edgesite/pkg/controller/config/kube.go]: Great comments. Thanks Kevin.

File: edgesite/pkg/controller/config/kube.go (hunk 3 above)
Comment (CindyXing) [reply edgesite/pkg/controller/config/kube.go]: This is the same as existing cloud side controller. I am keeping as is.

File: edgesite/pkg/controller/config/kube.go (hunk 6 above)
Comment (CindyXing) [reply edgesite/pkg/controller/config/kube.go]: KubeNamespace will be used by Edged. We are not changing the core code.

File: edgesite/pkg/controller/config/kube.go (hunk 7 above)
Comment (CindyXing) [reply edgesite/pkg/controller/config/kube.go]: Agree. This is true for edgecontroller. Let me raise a code clean up issue.

File: edgesite/pkg/controller/config/kube.go (hunk 8 above)
Comment (CindyXing) [reply edgesite/pkg/controller/config/kube.go]: Turn out github.com\go-chasis cast.value only supports ToFloat64. That was the reason.

File: edgesite/pkg/controller/config/kube.go (hunk 7 above)
Comment (kevin-wangzefeng) [reply edgesite/pkg/controller/config/kube.go]: Sure, please link the clean up issue to this PR, and we can get this one accepted.

**Comment by sids-b** [comment]
@kevin-wangzefeng @qizha @m1093782566 . The PR is ready for review

File: edgesite/pkg/controller/config/kube.go (hunk 7 above)
Comment (CindyXing) [reply edgesite/pkg/controller/config/kube.go]: Sounds good. #525 is raised for this purpose.

**Comment by rohitsardesai83** [comment]
Let's address the issue #531 post merging this. Thanks.

File: edgesite/conf/edgeSite.yaml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    image-gc-high-threshold: 80 # percent
+    image-gc-low-threshold: 40 # percent
+    maximum-dead-containers-per-container: 1
+    docker-address: unix:///var/run/docker.sock
+    version: 2.0.0
+
+metamanager:
+    context-send-group: controller
+    context-send-module: controller
+    edgesite: true
```
Comment (kevin-wangzefeng) [inline edgesite/conf/edgeSite.yaml]: No new line at end of file

**Comment by sids-b** [comment]
> LGTM.
> One none blocking issue is that we need version control for the component configs, which makes life easier for upgrades etc.. Let's clean it up in #525

Sure, Thanks

**Pull request merged**